Unique v,h

carterbox opened the pull request:
Adds the ability to have unique v, h positions at each view and fixes some bugs. This pull will cause the cause the API to change, so the version will be increased to v0.5.0.

dgursoy commented on tike/ptycho.py:
```diff
@@ -235,26 +236,22 @@ def combine_grids(grids, v, h,
+    combined = np.zeros([
+                         combined_shape[-2] + 2,
+                         combined_shape[-1] + 2], dtype=grids.dtype)
     # Add each of the grids to the appropriate place on the summed_grids
```
Squeeze 239-241 in 2 lines?

dgursoy commented on tike/ptycho.py:
```diff
@@ -178,28 +176,31 @@ def fast_pad(unpadded_grid, npadv, npadh):
     is faster because of the way that np.pad uses concatenate.
 
     """
+    assert npadv >= 0 and npadh >= 0, "Pads must be non-negative."
```
If pad is zero, then you can just return unpadded_grid.

dgursoy replied:
I think this depends on the size of the probe support. We may want to pad for simulations, because we use a small circular compact probe compared to the detector. But for real data, the probe can be larger like a sink shaped function with main intensity at the center with decaying tails. Therefore, It may make sense to set the probe shape  the same as the detector shape.

dgursoy commented on tike/ptycho.py:
```diff
@@ -178,28 +176,31 @@ def fast_pad(unpadded_grid, npadv, npadh):
 
 
 def combine_grids(grids, v, h,
                   combined_shape, combined_min):
```
Similar to above comment: combined_min to combined_corner?

dgursoy commented on tike/tike.py:
```diff
@@ -83,7 +83,7 @@ def _combined_interface(obj, obj_min,
+def admm(obj=None, voxelsize=1.0,
          data=None,
          probe=None, theta=None, h=None, v=None, energy=None,
          niter=1, rho=0.5, gamma=0.25,
```
Are the args explained in func docs?

carterbox replied:
Good catch. I am adding a docstring for `voxelsize` and `niter`. `rho`, `gamma` are actually parameters for `ptycho.grad`. Those need to be moved into `pkwargs` when we do the thing where we reduce the number of function parameters by packing everything into dictionaries.

dgursoy replied:
```rho``` is actually an ADMM parameter. On the ptychography side its default should to zero.

carterbox replied:
`rho` cannot be zero for `pytcho.grad`. That would cause a divide by zero error on [this line](https://github.com/carterbox/tike/blob/92c8a639670dc0bdd6e30eb5175a26df9cc57a52/tike/ptycho.py#L388-L391). It could be set to 1 by default?

dgursoy replied:
I'd default it with a very small number instead of one. 1e-12?

carterbox replied:
We can rearrange the math to prevent division by zero:
```python
+ gamma * rho * (reg - lamda / rho) # this is the original
+ gamma * (reg * rho - lamda) # this is rho multiplied through
```
@slnsrydn, said it was allowed.

dgursoy replied:
Yes. This is the way to do it.

carterbox commented on tike/ptycho.py:
```diff
@@ -215,8 +216,8 @@ def combine_grids(grids, v, h,
 
     # SHARED
```
Wouldn't this increase the complexity of the code and make it harder to read?

dgursoy replied:
If only one or two lines of code, maybe yes, but if more than >5 lines then it is debatable. It may desire a better docs about what is going there.

carterbox replied:
Yeah. Duplicate code was removed, but the net change in lines is zero because documentation.

carterbox commented on tike/ptycho.py:
```diff
@@ -95,9 +95,9 @@
 
 
 def _ptycho_interface(data, data_min,
```
I agree that this variable naming system is slightly ambiguous. However, this naming system is explained in the module level docstring. A clearer name would be `data_min_corner`, and I would be happy to adopt this naming system instead. However, you have often told me that short variable names are better, so that's why I chose `data_min` over `data_min_corner`.

Here is why I didn't choose any of your suggested alternatives:
- `axis_min`, `det_min` The corners of multiple things need to be specified, so these options are more confusing than the current option because they remove the connection implied by using the word `data` in both variable names.
- `data_corner` is probably equally ambiguous because then you need to ask "which corner?" There 4 options. However, `data_min` is shorter.

dgursoy replied:
I'd be more inclined towards ```data_corner``` and explaining that it is the minimum. My second best is ```data_min_corner```. 

As soon as all corners are consistent, it doesn't matter which corner is picked. So, maybe we find an automated or safe way for this. How about a correction with the minimum values of h and v of the whole stack?

carterbox replied:
> As soon as all corners are consistent, it doesn't matter which corner is picked.

This is not true. The alignment algorithms in `pytcho.combine_grids` rely on knowing that the corner is the min corner. Also, there is no way of determining which corner has been specified from the coordinates of the corners themselves.

I oppose the ability choose your own corner. This would add unneeded complexity to the code base when preprocessing the data to conform to our min corner specification is trivial.

dgursoy replied:
I would advocate that the quote still is true. Assume a grid of size (40, 60) and when you pick corners as (0, 0), (0, 20), (20, 0) or (20, 20) for min is zero. If I used a different corner these points will be (40, 60), (40, 80), (60, 60) or (60, 80). Then you subtract min(h-axis)=40 and min(v-axis)=60 from each axis.

I agree with you that this will add some unnecessary complexity, but also it will reduce the complexity in naming. It's your choice, but ```corner``` is a good thing to have in the variable.

carterbox replied:
Two corners are required to define a rectangle. If I only tell you the side lengths and the location of one corner, you cannot draw a box correctly. It's impossible for you to know which corner I have given you without more information.

I agree that having `corner` in the variable name is good. I'm only debating whether corner naming consistency is enough information to correctly arrange boxes.

dgursoy replied:
Wait a minute.. I'm confused. Why do we need this variable at all? Isn't ```h``` and ```v``` are defined on the global coordinates?

carterbox replied:
We need `data_corner` if the probe movement is independent of the detector, which it could be, but right now that feature is not implemented.

dgursoy replied:
Detector's are always fixed. So it is always the sample (or the probe in relative to the object) is moving. So, we may never need this argument.

dgursoy replied:
Probe movement will be on the order of 10s of microns at most, and your detector size 10s of centimeters. I think it is still on the negligible range.

carterbox replied:
Good to know! I've removed that parameter from all public functions, so approve this pull please.

carterbox commented:
I have addressed all of the comments relevant to this pull request and some additionally ones that were not in its original scope. The remaining unresolved comments are API breaking changes that I believe are not necessary at this time.
